### PR TITLE
[WIP] Refactor the exception management

### PR DESCRIPTION
**Pull request by Ener-Getick**
This PR attempts to refactor the exception management to allow the user to change the serialized output.

### CODE REVIEW

**Comment by Ener-Getick** [comment]
Another solution would be to add normalizers on top of the serializers (and not only for the exceptions) to avoid having different output for common classes...

File: Normalizer/ExceptionNormalizerInterface.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Normalizes an exception.
+ *
+ * @author Ener-Getick <egetick@gmail.com>
+ */
+interface ExceptionNormalizerInterface
+{
+    /**
+     * Normalizes an exception into a set of arrays/scalars.
+     public function normalize($exception);
```
Comment (stloyd) [inline Normalizer/ExceptionNormalizerInterface.php]: Missing `*`.

File: Normalizer/SimpleExceptionNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     * @param object $exception
+     *
+     * @return int
+     */
+    private function getExceptionStatusCode($exception)
+    {
+        if ($code = $this->isSubclassOf($exception, $this->codesMap)) {
+            return $code;
+        } elseif ($exception instanceof HttpExceptionInterface) {
```
Comment (stloyd) [inline Normalizer/SimpleExceptionNormalizer.php]: Should be just `if` as you call `return` in `if` above.

File: Normalizer/SimpleExceptionNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            return $exception->getStatusCode();
+        } else {
```
Comment (stloyd) [inline Normalizer/SimpleExceptionNormalizer.php]: `else` is useless as you call `return` in `if`.

File: Normalizer/SimpleExceptionNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     * @param object $exception
+     *
+     * @return array
+     */
+    private function getExceptionHeaders($exception)
+    {
+        if ($exception instanceof HttpExceptionInterface) {
+            return $exception->getHeaders();
+        } else {
```
Comment (stloyd) [inline Normalizer/SimpleExceptionNormalizer.php]: Same here.

File: EventListener/ExceptionListener.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function duplicateRequest(\Exception $exception, Request $request)
+    {
+        $attributes = array(
+           '_controller' => $this->controller,
+           'exception' => $exception,
+       );
```
Comment (stloyd) [inline EventListener/ExceptionListener.php]: Wrong indent.

**Comment by Ener-Getick** [comment]
Thanks @stloyd but this is not a definitive api. I'm wondering if we should not manage this differently on top of the new serializer adapters introduced by @xabbuh in 651714c94a7c2c61dcc83cff7e7d0410714253fa.

**Comment by lsmith77** [comment]
can you also look into incorporating the changes I did here https://github.com/FriendsOfSymfony/FOSRestBundle/commit/93758654c0ac2ed29576c0dcbe7d42122d71bf37 ?

**Comment by Ener-Getick** [comment]
I'm not sure about this PR.... might be better to provide different normalizers for each serializer. 
I'll search if we can provide a standard normalizer which will be adapted for each serializer then.

**Pull request closed without merging**